### ISSUE TITLE
[Feature] Allow for using "downlevelIteration" when downleveling

### ISSUE THREAD

**Issue by JoostK** [dherges/ng-packagr#418]
## Type of Issue

```
[ ] Bug Report
[X] Feature Request
```

## Description

I'd like to compile a library that depends on `downlevelIteration` to be enabled, a compiler flag that was introduced in TS 2.3. This is related to #256 but affects the options that are used during [downleveling ES2015 to ES5](https://github.com/dherges/ng-packagr/blob/be01602c673889019f5e81fd1157daa3699f0d39/src/lib/steps/tsc.ts#L22-L28), given that these options can not currently be changed. I had used a patched `tsconfig.ngc.json` before in which I enabled `downlevelIteration`, however that only affects the ES2015 build, and not the ES5 build for which the option actually matters ;)

#### Expected Behaviour

Possibly, we may be enabled to specify (some of) the TS compiler flags. This is already being discussed in #256 but would only affect ES2015 compilation, whereas `downlevelIteration` only affects the ES5 downleveling step that is hardcoded at the moment.

#### Version Information

```
ng-packagr: 2.0.0-rc.7
```

**Comment by JoostK** [comment]
Once again, I can whip up a PR but only after this week and once we have agreed upon a workable solution for this situation.

**Comment by dherges** [comment]
Is it safe to set `downLevelIteration` compiler flag, when down-levelling the ESM2015 bundle to a ESM5 bundle?

If yes, I'd say we should simply hard-code that compiler flag to the down-levelling build step.

**Comment by dherges** [comment]
By safe I mean: has `downlevelIteration` no effects other than `for ... of` language constructs?

The resulting es5 code is guaranteed to produce the expected behaviour in es5. Is there a use case for `downlevelIteration` in es2015 target?

**Comment by JoostK** [comment]
I thought about that question as well, and I'm not quite sure why TS decided to introduce a flag for this that is disabled by default.

I think the flag only affects ES5 generation. It does have influence on more stuff than just `for ... of`, e.g.

```
const map = new Map();
const values = [...map.values()]
```

Without `downlevelIteration`:
```
const map = new Map();
const values = map.values().slice();
```

With `downlevelIteration`:
```
const map = new Map();
const values = __spread(map.values());
```

The former is incorrect as Map.values returns an iterable which does not have `slice`. I don't think ES2015 code would be affected by this as syntax can just be preserved.

_What I think is odd is that the Angular CLI app that consumed the generated library was using the `.slice` notation, even though I would suspect it to just use the ES2015 bundle instead of the ES5 bundle when producing Webpack bundles_. While typing this I realise what is actually going on. My library consumers did not have the `downlevelIteration` flag enabled and would therefore generate `.slice` calls instead of `__spread` when downleveling all bundled code using Webpack. So, this issue is only relevant when using the UMD ES5 bundle, e.g. with SystemJS, but is otherwise not necessary as Angular CLI consumes ES2015 sources.

Nevertheless, I do not know what the effect of always enabling `downlevelIteration` exactly is, I presume it won't hurt anyone.

**Comment by JoostK** [comment]
Alright, so it turns out that my theory above was incorrect. Enabling `downlevelIteration` in the ES2015 > ES5 transformation does resolve the issue for library consumers. So, enabling the option in ng-packagr should indeed help.

I also had a further look into the origin of the flag and [the 2.3 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html) do clarify things. First, the flag only affects ES5 generation, so that's good. Furthermore, it is disabled by default as the TS team felt that only few people would need this support and because enabling it produces some more bytes, they decided to disable it by default.

**Comment by dherges** [comment]
Thanks for digging into that!

Imo, `downlevelIteration: true` is a common convention for the ES2015/ES5 downlevelling.

https://github.com/dherges/ng-packagr/blob/master/src/lib/steps/tsc.ts#L22

**Comment by dherges** [comment]
Solved by #475
